src/ts/spatialnavigation/navigationalgorithm.ts

```diff
@@ -0,0 +1,126 @@
+  const len = length(vector);
+
+  return {
+    x: vector.x / len,
+    y: vector.y / len,
+  };
+}
+
+/**
+ * Calculates the dot product between 2 vectors.
+ *
+ * @param a The first vector
+ * @param b The second vector
+ */
+function dotProduct(a: Vector, b: Vector): number {
+  return a.x * b.x + a.y * b.y;
+}
+
+/**
+ * Calculates the distance between the 2 points pointed to by the provided vectors.
+ *
+ * @param a The first vector
+ * @param b The second vector
+ */
+function distance(a: Vector, b: Vector): number {
+  return length({
+    x: b.x - a.x,
+    y: b.y - a.y,
+  });
+}
+
+/**
+ * Returns a vector that corresponds to the center of the provided element.
+ *
+ * @param element The element to get the center of
+ */
+function getElementVector(element: HTMLElement): Vector {
+  const boundingRect = element.getBoundingClientRect();
+
+  return {
+    x: boundingRect.x + boundingRect.width / 2,
+    y: boundingRect.y + boundingRect.height / 2,
+  };
+}
+
+/**
+ * Returns the angle in degrees between the unit vector pointing in the given {Direction} and the unit vector that
+ * points from the current element to another element.
+ *
+ * @param a The vector of the current element
+ * @param b The vector of the other element
+ * @param direction The direction to move along
+ */
+function calculateAngle(a: Vector, b: Vector, direction: Direction): number {
+  const directionVector = {
+    x: (direction === Direction.LEFT ? -1 : direction === Direction.RIGHT ? 1 : 0),
+    y: (direction === Direction.UP ? -1 : direction === Direction.DOWN ? 1 : 0),
+  };
+
+  const elementVector = normalize({
+    x: b.x - a.x,
+    y: b.y - a.y,
+  });
+
+  const angleCos = dotProduct(directionVector, elementVector) / (length(directionVector) * length(elementVector));
+
+  return Math.acos(angleCos) * 180 / Math.PI;
+}
+
+/**
+ * Returns the closest element to the current element when trying to navigate in the provided direction. Returns
+ * undefined, if there is not element in the given direction.
+ *
+ * @param activeElement The currently selected element
+ * @param elements The list of all elements that can be navigated to
+ * @param direction The direction in which to navigate
+ */
+export function getElementInDirection(
+  activeElement: HTMLElement,
+  elements: HTMLElement[],
+  direction: Direction,
+): HTMLElement | undefined {
+  if (!activeElement) return undefined;
+
+  const cutoffAngle = 45;
+  const activeElemVector = getElementVector(activeElement);
+
+  return elements
+    // don't take the current element into account
+    .filter(elem => elem !== activeElement)
+    // get the angle between, and distance to any other element from the current element
+    .map(elem => {
+      const elemVector = getElementVector(elem);
+      const dist = distance(activeElemVector, elemVector);
+      const angle = calculateAngle(activeElemVector, elemVector, direction);
+      return [angle, dist, elem] as const;
```

Comment:
Nit: would it make sense to use an object with named parameters here? 🤔 I think that would allow to use the name of the property in `.shift()?.[2];` instead of `2`

Reply:
Addressed in [e3e404c](https://github.com/bitmovin/bitmovin-player-ui/pull/468/commits/e3e404ce668e068d91d4800e0d6c707b433ebddf)